weave/trace_server/objects_query_builder.py

```diff
@@ -0,0 +1,275 @@
+from collections.abc import Iterator
+from typing import Any, Optional
+
+from weave.trace_server import trace_server_interface as tsi
+from weave.trace_server.clickhouse_schema import SelectableCHObjSchema
+from weave.trace_server.orm import combine_conditions
+from weave.trace_server.trace_server_common import digest_is_version_like
+
+VALID_OBJECT_SORT_FIELDS = {"created_at", "object_id"}
+VALID_SORT_DIRECTIONS = {"asc", "desc"}
+OBJECT_METADATA_COLUMNS = [
+    "project_id",
+    "object_id",
+    "created_at",
+    "refs",
+    "kind",
+    "base_object_class",
+    "digest",
+    "version_index",
+    "is_latest",
+    # columns not used in SelectableCHObjSchema():
+    "version_count",
+    "is_op",
+]
+
+
+def _make_optional_part(query_keyword: str, part: Optional[str]) -> str:
+    if part is None or part == "":
+        return ""
+    return f"{query_keyword} {part}"
+
+
+def _make_limit_part(limit: Optional[int]) -> str:
+    if limit is None:
+        return ""
+    return _make_optional_part("LIMIT", str(limit))
+
+
+def _make_offset_part(offset: Optional[int]) -> str:
+    if offset is None:
+        return ""
+    return _make_optional_part("OFFSET", str(offset))
+
+
+def _make_sort_part(sort_by: Optional[list[tsi.SortBy]]) -> str:
+    if not sort_by:
+        return ""
+
+    sort_clauses = []
+    for sort in sort_by:
+        if (
+            sort.field in VALID_OBJECT_SORT_FIELDS
+            and sort.direction in VALID_SORT_DIRECTIONS
+        ):
+            sort_clause = f"{sort.field} {sort.direction.upper()}"
+            sort_clauses.append(sort_clause)
+    return _make_optional_part("ORDER BY", ", ".join(sort_clauses))
+
+
+def _make_conditions_part(conditions: Optional[list[str]]) -> str:
+    if not conditions:
+        return ""
+    conditions_str = combine_conditions(conditions, "AND")
+    return _make_optional_part("WHERE", conditions_str)
+
+
+def _make_object_id_conditions_part(object_id_conditions: Optional[list[str]]) -> str:
+    if not object_id_conditions:
+        return ""
+    conditions_str = combine_conditions(object_id_conditions, "AND")
+    return _make_optional_part("AND", conditions_str)
+
+
+def format_metadata_objects_from_query_result(
+    query_result: Iterator[tuple[Any, ...]],
+) -> list[SelectableCHObjSchema]:
+    result = []
+    for row in query_result:
+        # Add an empty val_dump to the end of the row
+        row_with_val_dump = row + ("{}",)
+        columns_with_val_dump = OBJECT_METADATA_COLUMNS + ["val_dump"]
+        row_dict = dict(zip(columns_with_val_dump, row_with_val_dump))
+        row_model = SelectableCHObjSchema.model_validate(row_dict)
+        result.append(row_model)
+    return result
+
+
+class ObjectQueryBuilder:
```

Comment:
as authored, this is a "MetadataQueryBuilder" since you can't get the vals out

Reply:
Hmm true. I'll add the value query...

Reply:
ended up just renaming, the value query is so simple it doesn't really make sense to refactor